marlowe-playground-client/src/Marlowe/Linter.purs

```diff
@@ -138,198 +147,427 @@ termToRange a { row, column } =
   , endColumn: column + (length (show a))
   }
 
--- | We go through a contract term collecting all warnings and holes etc so that we can display them in the editor
--- | The aim here is to only traverse the contract once since we are concerned about performance with the linting
--- FIXME: There is a bug where if you create holes with the same name in different When blocks they are missing from
--- the final lint result. After debugging it's strange because they seem to exist in intermediate states.
-lint :: Term Contract -> State
-lint = go mempty
-  where
-  go :: State -> Term Contract -> State
-  go state (Term Close _) = state
-
-  go state (Term (Pay acc payee token payment contract) _) =
-    let
-      gatherHoles = getHoles acc <> getHoles payee <> getHoles token
-
-      newState =
-        state
-          # over _holes gatherHoles
-          # over _warnings (maybeInsert (NegativePayment <$> negativeValue payment))
-    in
-      go newState contract <> lintValue newState payment
-
-  go state (Term (If obs c1 c2) _) = go state c1 <> go state c2 <> lintObservation state obs
+newtype LintEnv
+  = LintEnv
+  { letBindings :: Set ValueId
+  , maxTimeout :: MaxTimeout
+  }
```

Comment:
I don't understand what `LintEnv` is for, why can't you just use `State` for these two fields?

Reply:
Because that way you get the type system to prevent errors from you. Things in `LintEnv` have more restrictions than things in `State` in the way they propagate (they don't go out of the scope)

Reply:
can you give an example? I'm confused.

Reply:
Yes, the field `letBindings` in only propagates inwards: adding an element to the set in `letBindings` should only affect subtrees, things outside of the current `Let` should not be affected (it is not shadowing if it is out of scope), whereas `holes` you want to accumulate all of them throughout the contract, and you want them to be in the `State` when you finish traversing.  So basically you combine `letBindings` and `holes` differently, if you have them all in `State` then how do you combine the state resulting from one subtree with the state resulting from another subtree? And in fact I think there was a bug that was related to this, you can see the code in `When` analysis became so much simpler and I can just use `traverse_`, whereas code before was really complicated.